core/Storage.h

```diff
@@ -1,136 +1,275 @@
 #pragma once
 
-#include <memory>
+#include <chrono>
 #include <istream>
+#include <map>
+#include <memory>
+#include <optional>
 #include <ostream>
 
-#include "io/ismrmrd_types.h"
+#include "IsmrmrdContextVariables.h"
 #include "io/adapt_struct.h"
+#include "io/ismrmrd_types.h"
 #include "io/primitives.h"
 #include <boost/date_time/posix_time/posix_time_config.hpp>
 #include <boost/iostreams/device/array.hpp>
 #include <boost/iterator/transform_iterator.hpp>
 
 namespace Gadgetron::Storage {
-    class StreamProvider {
-    public:
-        virtual ~StreamProvider() = default;
 
-        [[nodiscard]] virtual std::vector<std::string>
-        content(const std::string &subject, const std::string &key) const = 0;
+struct StorageItemTags {
+  public:
+    class Builder;
 
-        [[nodiscard]] virtual std::vector<char> fetch(const std::string &uuid) const = 0;
+    std::string subject;
+    std::optional<std::string> device;
+    std::optional<std::string> session;
+    std::optional<std::string> name;
+    std::multimap<std::string, std::string> custom_tags;
+};
+
+class StorageItemTags::Builder {
+  public:
+    Builder(std::string const& subject) { tags.subject = subject; }
+
+    Builder& with_device(std::string const& device) {
+        tags.device = device;
+        return *this;
+    }
+
+    Builder& with_session(std::string const& session) {
+        tags.session = session;
+        return *this;
+    }
+
+    Builder& with_name(std::string const& name) {
+        tags.name = name;
+        return *this;
+    }
+
+    Builder& with_custom_tag(std::string const& tag_name, std::string const& tag_value) {
+        tags.custom_tags.insert({tag_name, tag_value});
+        return *this;
+    }
+
+    StorageItemTags build() { return tags; }
+
+  private:
+    StorageItemTags tags;
```

Comment:
Not sure whether the preferred style is this or `tags_`. I see both in the codebase...

Reply:
I prefer `tags_` but its not consistent in the codebase as you say. So either way.